Sources/CMakeLists.txt

```diff
@@ -12,12 +12,29 @@
 ##
 ##===----------------------------------------------------------------------===##
 
-add_subdirectory(_FoundationCShims)
-
-# Disable the macro build on Windows until we can correctly build it for the host architecture
-if(NOT CMAKE_SYSTEM_NAME STREQUAL Windows)
-	add_subdirectory(FoundationMacros)
+include(ExternalProject)
+if(CMAKE_HOST_WIN32)
+  set(_SwiftFlags -DCMAKE_Swift_FLAGS="-use-ld=lld")
+else()
+  set(_SwiftFlags)
+endif()
+ExternalProject_Add(FoundationMacros
+  SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/FoundationMacros"
+  PREFIX "${CMAKE_BINARY_DIR}/_deps"
+  BINARY_DIR "macros"
+  CMAKE_ARGS
+    -DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}
+    ${_SwiftFlags}
+  INSTALL_COMMAND "")
```

Comment:
Could you elaborate on what this setting does specifically? Does this disable the install step for the macros?

Reply:
Yes, this is disabling the install step of the external build.

Reply:
Will this also prevent installing the shared library on linux? That currently works as expected and we should try not to regress that if possible

Reply:
There is no guarantee that the library would be usable - if I built the library on macOS, the MachO binary wouldn't run on Linux.

Reply:
I think I might be missing some details here so I apologize if I don't have the right interpretation - but to me it seems like this will in effect disable the macro build on all platforms since the install command is empty. It makes sense that building the macro on macOS and deploying to linux wouldn't work, but do we support building as that configuration today? Right now we have the macro successfully building for linux and being installed into the linux toolchain in swift-ci via our cmake build and I want to make sure we don't regress that with this change.

Reply:
I think that we do support that, e.g. for the static SDKs that we are building. The android work also is done as Windows -> Android, and I've done Windows -> Linux in the past as well. This means that we would be installing Windows executables into a Linux installation. In order to install the macros prebuilt into the toolchain, I think that we should be building just the macros manually for Linux and installing them into the toolchain.

Reply:
It will still run the build and produce the macros into the build directory. It won't copy them into the final install location of the library itself. We don't have a great distribution story for macros and cross-compiling at the moment. Let's say, for example, that I'm on a Mac trying to cross-compile to a Linux. The macros will build for macOS so that I can run them locally while building, while the library will build for Linux. It doesn't really make sense to include the macOS-built macros in the Linux SDK. If there are macros that we need to ship, we will need to build them separately for all platforms that have a toolchain that can use the SDK. Long-term, WASM might be interesting in this space so that we only build the macros for WASM and then ship a WASM runtime in the toolchain , but I don't think that's in any of the current cards.

An alternative is to set the install prefix of the EP_Add to add `-DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>` to the cmake args and have it install the executables to a temporary location in the Foundation build directory, but I don't see too much point to copying temporaries around unless we have to.

Reply:
Note that failure to pass `-D CMAKE_INSTALL_PREFIX=...` after such a change would also break the build as it will attempt to install to the root where the user may not have permission (or if running as root, could install over the system copy).

Reply:
How do the Stdlib macros that are installed into the toolchain already today work? IMO we should follow the same pattern here, do they support cross compiling in this manner?

Reply:
The stdlib macros are built multiply (for each standard library host). They are not used to build the standard library or run its tests. We currently have references of the macros in the tests, which require that the macros are built for build. We would need to do a second build of the macros for the host, which will also require building swift-syntax for the host and increase the overall build times. However, that seems like the right thing to do if we need to ship the macros as prebuilt binaries. This is likely going to be problematic as the SDK and toolchains can go out of sync and there is no ABI stability outside of Apple platforms though and so would be better to avoid IMO.

Reply:
@etcwilde @jmschonfeld - what do we want to do here? Introduce an option to build a second copy of the macros?

Reply:
If I knew how to install them without them conflicting with each other, I'd suggest a loop over a list of target triples passed into the ExternalProject invocations so that someone can list off what toolchains their SDK will work with, sort of like how the LLVM build allows you to specify which triples you want to build the runtimes for.

Reply:
That seems like a good deal of infrastructure that we will need to build up and share - much like `SwiftSupport.cmake`, especially while we don't have `install(...)` working for Swift targets.

Reply:
For clarity - do we officially support cross-compiling a linux toolchain or is that just something that you _can_ do but isn't something that's load bearing at the moment? As of today, we're able to successfully produce linux toolchains that contain `libFoundationMacros.so` and the macro can be used when using that toolchain. I think this change would remove that library from the toolchain which would be a step backwards on Linux, and I'd like to avoid making that regression unless we deem that producing this library in the current manner is breaking a load bearing, officially supported path.

As for the windows side, yeah I'm not sure whether it'd be best to build the macro for just the host (to use when building Foundation itself) or just the destination (to package into the created toolchain). I suppose starting with the former and hopefully supporting the latter in the future is a path forward that we can take here, it's just unfortunate that Windows toolchains would be missing out on these APIs like `#Predicate`. This sounds like a general problem that we'll likely need to solve quickly given that swift-testing will likely run into the same issue shortly, but their macros are much more critical to their API surface (cc @grynspan / @stmontgomery)

Reply:
On Windows, at least this is "easier": the only supported hosts are windows x86_64, arm64. Because we build the SDK for both of these platforms in CI, we could build the macros for both of the hosts.

The problem is when we build for Android, what are the supported hosts? Windows x86_64, arm64, Linux x86_64, arm64, Darwin arm64, FreeBSD arm64, x86_64? What about Haiku? We would need to have the SDKs for all of those platforms available to build the macros. This quickly leads to an exponential growth problem.

Reply:
That's a good argument in favour of Wasm macro plugins.

Reply:
@jmschonfeld btw, AFAIK, cross-compilation of the toolchain for non-Darwin targets is not officially supported. While there are some intrepid users (e.g. Facebook and at times myself) who do this, it is simply on them to figure out how to piece that together. I do try to make it easy to cross-compile, but this might be one of those things where incremental improvements are the way to address this.